README.rst
dominionstats
=============

This is the code behind http://councilroom.com. Feel free to fork it to make
it do interesting new things.

The project is currently run by Mike McCallister
(mike@mccllstr.com). The maintainer "emeritus" is rrenaud
(rrenaud@gmail.com), with contributions so far by Larry, rspeer, David
Lu, and tlstyer.


Setup for Local Development
===========================

This project now includes files to enable you to use Vagrant
(http://www.vagrantup.com) to create isolated environments for
development, testing, and production release builds. This is much
easier to recreate than trying to install all the pieces following the
Installation instructions below.

This Vagrant-based model splits activity into separate
environments. The first of these is your *development host*. The
development host is where high-level tools that manage the other
environments run. This is where you will checkout and edit the source
code. Unlike many models where development and testing is done on the
local machine, you will not build, install, or test the code that runs
the CouncilRoom.com site directly on your development host.

The next environment is the *development guest*. A development guest
is a temporary virtual machine instance where the application will
actually be installed and run. These environments will contain the
numerous pieces that are required to run the application in a tidy,
easy to recreate package.

To make this magic work, these instructions (and the corresponding
files in our source code repository) assume you will use the
VirtualBox *provider* in Vagrant. They further assume you will use the
Ansible *provisioner* in Vagrant. If you've gotten this far, you
probably already realize that Git is the version control system in
use.

So in order to play along, you will need to outfit your development
host with the necessary pieces. This is typically a one-time
activity.

Use the following steps on an Ubuntu machine:

#. Install the appropriate version of VirtualBox
   (https://www.virtualbox.org/) for your host machine.

#. Install the appropriate version of Vagrant
   (http://www.vagrantup.com) for your host machine.

#. Install Git and make sure it knows who you are::

       sudo apt-get install git
       git config --global user.name "Mike McCallister"
       git config --global user.email mike@mccllstr.com

#. Install the tools that let us create the Python virtual environment
   that some other Python-based tools, particularly Ansible, will run
   from within the development host::

       sudo apt-get install python-virtualenv python-pip

#. Install the packages that Ansible depends on, or else install the
   packages needed to build those dependencies in the Python
   virtualenv that is used to run Ansible.

   #. If you are in doubt, it is simplest and quickest just to install
      the pre-built packages with this command::

       sudo apt-get install python-paramiko python-jinja2 python-yaml

   #. If Ansible is incompatible with the packages available in your
      development host for some reason, install the build dependencies
      so it can be compiled by pip when it is installed in the virtual
      environment below::

       sudo apt-get build-dep python-crypto python-paramiko python-jinja2 python-yaml

      .. note:: This has not been specificically tested. Further work
         may be requrired to enable Ansible to build.

On a Windows PC, the necessary steps are:

#. TODO

With the above steps completed, you are ready to get a local copy of
the source code and launch the development guest. Use the following
steps:

#. Clone the source code repository to your machine. In this example,
   the source code will go into the directory named `dev`::

     git clone git@github.com:mikemccllstr/dominionstats.git dev

#. Switch to the repository clone and setup the virtual environment
   that will be needed for tools that run on the development host
   itself::

     cd dev
     virtualenv --system-site-packages .venv-devhost
     . .venv-devhost/bin/activate
     pip install -r requirements/devhost.txt

The above steps typically only need to be done once per local copy of
the source code. Once you have this done, each time you want to do
some development, you will repeat the following steps:

#. Activate the devhost virtual environment, launch the Vagrant
   instance, and ssh into it::

     cd dev
     . .venv-devhost/bin/activate
     vagrant up
     vagrant ssh

#. From within the Vagrant instance, you can import a day's worth of
   data and then launch the web application with the following
   commands::

     cd /vagrant
     virtualenv


Installation
====

Turbo frontend javascript hacking
====
super easy frontend modifications without any server setup::

     checkout code from github.
     google-chrome --allow-file-access-from-files
     browse to local html pages (eg, dominionstats/supply_win.html).
     edit local javascript files.

I just want access to some data!
====
Send (rrenaud@gmail.com) a mail and I'll see what I can do.

Long, arduous but incredibly rewarding full setup
====
The code depends on:

- Python version 2.6 or 2.7
- mongodb (http://www.mongodb.org) (1.5.3+ or later)
- pymongo (http://api.mongodb.org/python/1.9%2B/index.html)
- web.py (http://webpy.org)
- argparse (included in Python 2.7/3.2)
- simplejson (http://pypi.python.org/pypi/simplejson/)

Ubuntu Installation Commands for pymongo and webpy::

     sudo pip install pymongo
     sudo easy_install web.py
     mkdir db

Run an instance of mongodb with::

     mongod --dbpath=db

After install those packages, the system can be setup by running the
update_loop.py script, which will take a few hours to download one months of
games logs from councilroom, and then parse through it all and load it into a
database::

     python update_loop.py
These instructions are outdated and no longer work. Use update.py, which is run through a worker script.


And after that is down, this starts webserver running on localhost:8080::

     python frontend_local.py

Hacking Guidelines
====
Python code:
  - Write it in pep8, even if I didn't all the time.
  - Wrap lines at 80 characters.
  - Try not to write super long functions, break them up into logical subfunction even if those functions are only called once.

  - Did you see something in the code is fugly and offends your natural sense of what is good in the world?  I'll happily take style cleanups.

  - Anatomy of an analysis.
     + Want analysis to be incremental, can go day at a time.
        * Use an incremental_scanner.
        * Prefer using game.Game objects over raw game docs from the database.  game.Game objects are easier to work with, and make maintenance easier.
        * Since we want to be incremental, store raw counts in database, normalize them with divisions, etc at presentation time.
        * Want to keep track of a random variable that has some kind of spread? Use a stats.MeanVarStat.
        * Consider using primitive_util for serializing/deserializing objects that store the aggregate information.

  - How to do display?
      + Lots of existing server side templating in Python.
      + Prefer templates to programatically building up strings (even if if the existing code doesn't always).
      + Seriously consider outputting JSON from server like (/supply_win_api) and doing presentation in javascript (/supply_win), since it allows frontend development without running the whole system.

JavaScript code:
  * Keep non-trivial bits of JavaScript in separate .js files rather than embedded in HTML.  It makes my Emacs happier.

Got a technical problem/question/idea?  You can send an email to the dev group,

https://groups.google.com/forum/?fromgroups#!forum/councilroom-dev

Happy hacking.


debian/changelog
councilroom (1.0.0) unstable; urgency=low

  * Initial Release.

 -- Mike McCallister <mike@mccllstr.com>  Sun, 04 Aug 2013 21:23:03 +0000


utils.py
#!/usr/bin/python

from datetime import timedelta, datetime, date
import ConfigParser
import argparse
import logging
import os
import pymongo
import time

import primitive_util


# Module-level logging instance
log = logging.getLogger(__name__)
log.addHandler(logging.NullHandler())


# http://stackoverflow.com/questions/312443/how-do-you-split-a-list-into-evenly-sized-chunks-in-python
def segments(lis, chunk_size):
    """ Return an iterator over sublists whose size matches chunk_size. """
    for i in xrange(0, len(lis), chunk_size):
        yield lis[i:i + chunk_size]


def get_aws_credentials():
    """Retrieve the AWS credentials from the config file

    Returned in a dict, so you can make a call like:

        boto.s3.connection.S3Connection(**utils.get_aws_credentials())
    """
    config = ConfigParser.ConfigParser()
    config.read('conf.ini')

    return {'aws_access_key_id': config.get('aws', 'aws_access_key_id'),
            'aws_secret_access_key': config.get('aws', 'aws_secret_access_key')}


def get_mongo_connection():
   mongo_connection = 'localhost'

   try:
      config = ConfigParser.ConfigParser()
      config.read('conf.ini')

      mongo_connection = config.get('mongo', 'connection')
   #FIXME: this is too broad
   except:
      pass

   return pymongo.Connection(mongo_connection)

# Should I read once somewhere and cache?  I guess when
#   we have more config things.
def get_mongo_database():
   connection = get_mongo_connection()

   db = None
   try:
      config = ConfigParser.ConfigParser()
      config.read('conf.ini')

      db = connection[config.get('mongo', 'database')]
   except:
      db = connection['test']

   return db


def get_bad_leaderboard_dates():
    """Return a list of leaderboard dates that should be skipped

    List comes from the conf.ini file, as a multi-line entry under:

    [leaderboard]
    known bad dates = 2011-11-24
        2011-11-25
        2011-11-26
        2011-11-27
        2011-11-28
    """

    try:
        config = ConfigParser.ConfigParser()
        config.read('conf.ini')
        bad_dates = str(config.get('leaderboard', 'known bad dates')).strip().splitlines()
    except:
        log.exception("Got exception, using default list")
        bad_dates = ['2011-11-24', '2011-11-25', '2011-11-26', '2011-11-27',
                     '2011-11-28', '2011-11-29', '2011-11-30', '2011-12-01',
                     '2011-12-02', '2011-12-03', '2011-12-04', '2012-06-08', ]

    return bad_dates


def read_object_from_db(obj, collection, _id):
   prim = collection.find_one({'_id': _id})
   if prim:
      obj.from_primitive_object(prim)

def write_object_to_db(obj, collection, _id):
    prim = primitive_util.to_primitive(obj)
    prim['_id'] = _id
    collection.save(prim)

def ensure_exists(path):
    if not os.path.exists(path):
        os.makedirs(path)

def at_least_as_big_as(path, min_file_size):
    if not os.path.exists(path):
        return False
    return os.stat(path).st_size >= min_file_size


def daterange(start_date, end_date, reverse=False):
    """Returns a generator that produces the datetime.date() objects
    that are between start_date (inclusive) and end_date (exclusive).

    Normally returns dates in the natural order between start_date and
    end_date, but when reverse=True is passed, the sequence will be
    reversed.
    """
    if end_date >= start_date:
        step = 1
    else:
        step = -1

    sequence = range(0, (end_date - start_date).days, step)
    if reverse:
        sequence = reversed(sequence)

    for n in sequence:
        yield start_date + timedelta(n)


def base_parser():
    """Root command line parser for scripts"""
    parser = argparse.ArgumentParser()
    parser.add_argument('--debug', dest='debug', action='store_true',
                        default=False, help='Output debug detail')
    return parser

def incremental_parser():
    """Command line parser for scripts that handle the --noincremental arg"""
    parser = base_parser()
    parser.add_argument('--noincremental', action='store_false',
                        dest='incremental')
    return parser

def incremental_max_parser():
   parser = incremental_parser()
   parser.add_argument('--max_games', default=-1, type=int)
   return parser

def date_arg_convert(string_input):
    """ Converts date representation on the command line into a datetime.date
    """
    if string_input == 'yesterday':
        return date.today() - timedelta(days=1)
    elif string_input.startswith('days-ago'):
        # Handle arg in format "days-ago-3"
        try:
            days_delta = int(string_input[9:])
            return date.today() - timedelta(days=days_delta)
        except ValueError:
            raise argparse.ArgumentTypeError("The 'days-ago-##' syntax requires an integer")
    else:
        try:
            # Try parsing from ISO format
            return datetime.strptime(string_input, "%Y-%m-%d").date()
        except ValueError:
            raise argparse.ArgumentTypeError("Date value '%s' not in ISO 8601 format" % string_input)

def add_date_range_args(parser):
    # 20101015 is the first day with standard turn labels
    parser.add_argument('--start-date', dest='startdate',
                        default=date(2010, 10, 15),
                        type=date_arg_convert)
    parser.add_argument('--end-date', dest='enddate',
                        default=date.today(),
                        type=date_arg_convert)
    return parser

def incremental_date_range_cmd_line_parser():
    parser = add_date_range_args(incremental_parser())
    parser.add_argument('--passive', action='store_true')
    return parser

def includes_day(args, str_yyyymmdd):
    assert len(str_yyyymmdd) == 8, '%s not 8 chars' % str_yyyymmdd
    return args.startdate <= str_yyyymmdd <= args.enddate

def progress_meter(iterable, passed_log=None, chunksize=1000):
    """Logs progress through iterable at chunksize intervals.

    Note, passed_log is now ignored.
    """
    scan_start = time.time()
    since_last = time.time()
    for idx, val in enumerate(iterable):
        if idx % chunksize == 0 and idx > 0:
            log.info("Iteration: %5d; avg rate: %7.1f/s; inst rate: %7.1f/s",
                     idx,
                     idx / (time.time() - scan_start),
                     chunksize / (time.time() - since_last))
            since_last = time.time()
        yield val


# See http://code.activestate.com/recipes/498245-lru-and-lfu-cache-decorators/
import collections
import functools

def lru_cache(maxsize=100):
    '''Least-recently-used cache decorator.

    Arguments to the cached function must be hashable.
    Cache performance statistics stored in f.hits and f.misses.
    http://en.wikipedia.org/wiki/Cache_algorithms#Least_Recently_Used

    '''
    def decorating_function(user_function):
        cache = collections.OrderedDict()    # order: least recent to most recent

        @functools.wraps(user_function)
        def wrapper(*args, **kwds):
            key = args
            if kwds:
                key += tuple(sorted(kwds.items()))
            try:
                result = cache.pop(key)
                wrapper.hits += 1
            except KeyError:
                result = user_function(*args, **kwds)
                wrapper.misses += 1
                if len(cache) >= maxsize:
                    cache.popitem(0)    # purge least recently used cache entry
            cache[key] = result         # record recent use of this key
            return result
        wrapper.hits = wrapper.misses = 0
        return wrapper
    return decorating_function
